crates/wasmtime/src/runtime/gc/enabled.rs

```diff
@@ -0,0 +1,391 @@
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn _foo() -> Result<()> {
+    /// let mut store = Store::<()>::default();
+    ///
+    /// let externref = ExternRef::new::<usize>(&mut store, 0);
+    ///
+    /// // Access the `externref`'s host data.
+    /// let data = externref.data_mut(&mut store)?;
+    /// // Dowcast it to a `usize`.
+    /// let data = data.downcast_mut::<usize>().ok_or_else(|| Error::msg("not a usize"))?;
+    /// // We initialized to zero.
+    /// assert_eq!(*data, 0);
+    /// // And we can mutate the value!
+    /// *data += 10;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn data_mut<'a, T>(
+        &self,
+        store: impl Into<StoreContextMut<'a, T>>,
+    ) -> Result<&'a mut (dyn Any + Send + Sync)>
+    where
+        T: 'a,
+    {
+        let store = store.into();
+
+        // Safety: we don't do anything that could cause a GC while handling
+        // this `gc_ref`.
+        //
+        // NB: We can't use AutoAssertNoGc` here because then the lifetime of
+        // `gc_ref.as_extern_ref()` would only be the lifetime of the `store`
+        // local, rather than `'a`.
+        let gc_ref = unsafe { self.inner.unchecked_try_gc_ref_mut(store.0)? };
+
+        let externref = gc_ref.as_extern_ref_mut();
+        // Safety: We have a mutable borrow on the store, which prevents
+        // concurrent access to the underlying `VMExternRef`.
+        Ok(unsafe { externref.data_mut() })
+    }
+
+    /// Creates a new strongly-owned [`ExternRef`] from the raw value provided.
+    ///
+    /// This is intended to be used in conjunction with [`Func::new_unchecked`],
+    /// [`Func::call_unchecked`], and [`ValRaw`] with its `externref` field.
+    ///
+    /// This function assumes that `raw` is an externref value which is
+    /// currently rooted within the [`Store`].
+    ///
+    /// # Unsafety
+    ///
+    /// This function is particularly `unsafe` because `raw` not only must be a
+    /// valid externref value produced prior by `to_raw` but it must also be
+    /// correctly rooted within the store. When arguments are provided to a
+    /// callback with [`Func::new_unchecked`], for example, or returned via
+    /// [`Func::call_unchecked`], if a GC is performed within the store then
+    /// floating externref values are not rooted and will be GC'd, meaning that
+    /// this function will no longer be safe to call with the values cleaned up.
+    /// This function must be invoked *before* possible GC operations can happen
+    /// (such as calling wasm).
+    ///
+    /// When in doubt try to not use this. Instead use the safe Rust APIs of
+    /// [`TypedFunc`] and friends.
+    ///
+    /// [`Func::call_unchecked`]: crate::Func::call_unchecked
+    /// [`Func::new_unchecked`]: crate::Func::new_unchecked
+    /// [`Store`]: crate::Store
+    /// [`TypedFunc`]: crate::TypedFunc
+    /// [`ValRaw`]: crate::ValRaw
+    pub unsafe fn from_raw(
```

Comment:
Is it actually even possible to safely call this method? 

This only exist for the C API, and there's no rooting in the C API, so thinking about this it seems like we should remove this because it basically can't be used correctly

Reply:
> This only exist for the C API

We use this to convert `ValRaw`s back to `ExternRef`s, and the untyped function call path internally uses `ValRaw`s, so we can't just remove this without affecting non-C-API things.

> there's no rooting in the C API

There isn't explicit control of rooting, no, I haven't added C-API stuff for `RootScope` and `Rooted` vs `ManuallyRooted`. But we do root things in the C-API implementation nonetheless, for example a `wasm_ref_t` is a `ManuallyRooted<Ref>` under the hood and you could use this method safely in combination with that type.

Reply:
Oh right yeah, but the untyped/unchecked path basically only exists for the C API. I'm definitely not saying that we should add a bunch of rooting-related things to the C API, but I'm wondering if we should basically forbid externref in the C API as a result of this (or at least in the `ValRaw` parts bound in the C API). 

This seems like it might be too big of a footgun to expose in the C API where you can very easily get an unrooted value and there's no way to go from that unrooted value to a rooted value, so even if you want to "do the right thing" you can't

Reply:
I think it does make sense to expose the rooting stuff in the C API eventually, it's just not a priority at the moment. Given that, I think we can forbid it in the C API in the meantime, yeah. How would you like to do this? Propagating errors or asserting? I think the former could be difficult/annoying because of the amount of places that are currently infallible but which would need to become fallible...

Reply:
Perhaps removing the `externref` field from the header (and the tag for that enum)? And then using `unreachable!()` everywhere inside of the C bindings since they shouldn't be triggerable?

Reply:
Done.